libs/lib-postgres/src/db/connection/DatabaseClient.ts

```diff
@@ -0,0 +1,250 @@
+   */
+  schema?: string;
+  /**
+   * Notification channels to listen to.
+   */
+  notificationChannels?: string[];
+};
+
+export type DatabaseClientListener = NotificationListener & {
+  connectionCreated?: (connection: pgwire.PgConnection) => Promise<void>;
+};
+
+export const TRANSACTION_CONNECTION_COUNT = 5;
+
+export class DatabaseClient extends AbstractPostgresConnection<DatabaseClientListener> {
+  closed: boolean;
+
+  protected pool: pgwire.PgClient;
+  protected connections: ConnectionSlot[];
+
+  protected initialized: Promise<void>;
+  protected queue: DeferredPromise<ConnectionLease>[];
+
+  constructor(protected options: DatabaseClientOptions) {
+    super();
+    this.closed = false;
+    this.pool = pgwire.connectPgWirePool(options.config);
+    this.connections = Array.from({ length: TRANSACTION_CONNECTION_COUNT }, () => {
+      const slot = new ConnectionSlot({ config: options.config, notificationChannels: options.notificationChannels });
+      slot.registerListener({
+        connectionAvailable: () => this.processConnectionQueue(),
+        connectionError: (ex) => this.handleConnectionError(ex),
+        connectionCreated: (connection) => this.iterateAsyncListeners(async (l) => l.connectionCreated?.(connection))
+      });
+      return slot;
+    });
+    this.queue = [];
+    this.initialized = this.initialize();
+  }
+
+  protected get baseConnection() {
+    return this.pool;
+  }
+
+  protected get schemaStatement() {
+    const { schema } = this.options;
+    if (!schema) {
+      return;
+    }
+    return {
+      statement: `SET search_path TO ${schema};`
+    };
+  }
+
+  registerListener(listener: Partial<DatabaseClientListener>): () => void {
+    let disposeNotification: (() => void) | null = null;
+    if ('notification' in listener) {
+      // Pass this on to the first connection slot
+      // It will only actively listen on the connection once a listener has been registered
+      disposeNotification = this.connections[0].registerListener({
+        notification: listener.notification
+      });
+      delete listener['notification'];
+    }
+
+    const superDispose = super.registerListener(listener);
+    return () => {
+      disposeNotification?.();
+      superDispose();
+    };
+  }
+
+  /**
+   * There is no direct way to set the default schema with pgwire.
+   * This hack uses multiple statements in order to always ensure the
+   * appropriate connection uses the correct schema.
+   */
+  query(script: string, options?: pgwire.PgSimpleQueryOptions): Promise<pgwire.PgResult>;
+  query(...args: pgwire.Statement[]): Promise<pgwire.PgResult>;
+  async query(...args: any[]): Promise<pgwire.PgResult> {
+    await this.initialized;
+
+    const { schemaStatement } = this;
+    if (typeof args[0] == 'object' && schemaStatement) {
+      args.unshift(schemaStatement);
+    } else if (typeof args[0] == 'string' && schemaStatement) {
+      args[0] = `${schemaStatement.statement}; ${args[0]}`;
+    }
```

Comment:
It would be great if we could run the schemaStatement once per connection, instead of on every query. That's probably tricky with the pgwrire connection pool though.

Perhaps just add a comment explaining what is happening here for now?

Reply:
Yeah it is tricky to implement this once with the connection pool. I've moved the method comment closer to this implementation. 
The connection slots, which provide access to a single connection for a period of time only set this once.